### ISSUE TITLE
uBridge issue on Cloud interfaces

### ISSUE THREAD

**Issue by Ventresca68** [GNS3/gns3-server#1730]
Hello, I'm experiencing this behaviour on a recently updated 2.2.6 version.

When I try to connect a TAP interface insede a cloud with an ethernet IOU L3 interface I get this error:
```
uBridge process has stopped, return code: 1
uBridge version 0.9.17 running with libpcap version 1.8.1
Hypervisor TCP control server started (IP localhost port 36769).
UDP tunnel connecting from local port 10001 to IPv4 address 127.0.0.1 on port 10000
Source NIO listener thread for b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 has started
Destination NIO listener thread for b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 has started
recv: Connection refused
Source NIO listener thread for b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 has stopped because of an error: Invalid argument 
```
I've debugged a bit and I've found that when uBridge starts the tap0-udp bridge it gets a "connection refused" from the other side.

This is the relevant part of the debug:
```
2020-03-26 15:59:38 DEBUG compute.py:509 Attempting request to compute: POST http://localhost:3080/v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/ports/udp {'content-type': 'application/json'}
2020-03-26 15:59:38 DEBUG port_manager.py:288 UDP port 10000 has been allocated
2020-03-26 15:59:38 INFO response.py:58 POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/ports/udp
2020-03-26 15:59:38 DEBUG response.py:59 {'Host': 'localhost:3080', 'Content-Type': 'application/json', 'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Python/3.6 aiohttp/3.6.2', 'Authorization': 'Basic YWRtaW46Z3dydlFwekJqRXl4bFdiQ1NhTVhsQmR0R2pPbU9LQno3b0RrbmdhU3dGZGk1dEZFZndYbDZ2azAxUmdXMzNBNw==', 'Content-Length': '0'}
2020-03-26 15:59:38 DEBUG response.py:61 {}
2020-03-26 15:59:38 INFO response.py:62 Response: 201 Created
2020-03-26 15:59:38 DEBUG response.py:63 {'Connection': 'close', 'X-Route': '/v2/compute/projects/{project_id}/ports/udp', 'Server': 'Python/3.6 GNS3/2.2.6', 'Content-Type': 'application/json'}
2020-03-26 15:59:38 DEBUG response.py:65 {'udp_port': 10000}
2020-03-26 15:59:38 INFO web_log.py:233 ::1 [26/Mar/2020:14:59:38 +0000] "POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/ports/udp HTTP/1.1" 201 242 "-" "Python/3.6 aiohttp/3.6.2"
2020-03-26 15:59:38 DEBUG compute.py:509 Attempting request to compute: POST http://localhost:3080/v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/ports/udp {'content-type': 'application/json'}
2020-03-26 15:59:38 DEBUG port_manager.py:288 UDP port 10001 has been allocated
2020-03-26 15:59:38 INFO response.py:58 POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/ports/udp
2020-03-26 15:59:38 DEBUG response.py:59 {'Host': 'localhost:3080', 'Content-Type': 'application/json', 'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Python/3.6 aiohttp/3.6.2', 'Authorization': 'Basic YWRtaW46Z3dydlFwekJqRXl4bFdiQ1NhTVhsQmR0R2pPbU9LQno3b0RrbmdhU3dGZGk1dEZFZndYbDZ2azAxUmdXMzNBNw==', 'Content-Length': '0'}
2020-03-26 15:59:38 DEBUG response.py:61 {}
2020-03-26 15:59:38 INFO response.py:62 Response: 201 Created
2020-03-26 15:59:38 DEBUG response.py:63 {'Connection': 'close', 'X-Route': '/v2/compute/projects/{project_id}/ports/udp', 'Server': 'Python/3.6 GNS3/2.2.6', 'Content-Type': 'application/json'}
2020-03-26 15:59:38 DEBUG response.py:65 {'udp_port': 10001}
2020-03-26 15:59:38 INFO web_log.py:233 127.0.0.1 [26/Mar/2020:14:59:38 +0000] "POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/ports/udp HTTP/1.1" 201 242 "-" "Python/3.6 aiohttp/3.6.2"
2020-03-26 15:59:38 DEBUG compute.py:509 Attempting request to compute: POST http://localhost:3080/v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/iou/nodes/18f5c42d-80f6-4fac-9886-242632e81901/adapters/0/ports/0/nio {'content-type': 'application/json'}
2020-03-26 15:59:38 INFO iou_vm.py:906 IOU "IOU1" [18f5c42d-80f6-4fac-9886-242632e81901]: NIO UDP added to 0/0
2020-03-26 15:59:38 INFO response.py:58 POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/iou/nodes/18f5c42d-80f6-4fac-9886-242632e81901/adapters/0/ports/0/nio
2020-03-26 15:59:38 DEBUG response.py:59 {'Host': 'localhost:3080', 'Content-Type': 'application/json', 'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Python/3.6 aiohttp/3.6.2', 'Authorization': 'Basic YWRtaW46Z3dydlFwekJqRXl4bFdiQ1NhTVhsQmR0R2pPbU9LQno3b0RrbmdhU3dGZGk1dEZFZndYbDZ2azAxUmdXMzNBNw==', 'Content-Length': '106'}
2020-03-26 15:59:38 DEBUG response.py:61 {'lport': 10000, 'rhost': '127.0.0.1', 'rport': 10001, 'type': 'nio_udp', 'filters': {}, 'suspend': False}
2020-03-26 15:59:38 INFO response.py:62 Response: 201 Created
2020-03-26 15:59:38 DEBUG response.py:63 {'Connection': 'close', 'X-Route': '/v2/compute/projects/{project_id}/iou/nodes/{node_id}/adapters/{adapter_number:\\d+}/ports/{port_number:\\d+}/nio', 'Server': 'Python/3.6 GNS3/2.2.6', 'Content-Type': 'application/json'}
2020-03-26 15:59:38 DEBUG response.py:65 {'lport': 10000, 'rhost': '127.0.0.1', 'rport': 10001, 'type': 'nio_udp'}
2020-03-26 15:59:38 INFO web_log.py:233 ::1 [26/Mar/2020:14:59:38 +0000] "POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/iou/nodes/18f5c42d-80f6-4fac-9886-242632e81901/adapters/0/ports/0/nio HTTP/1.1" 201 376 "-" "Python/3.6 aiohttp/3.6.2"
2020-03-26 15:59:38 DEBUG compute.py:509 Attempting request to compute: POST http://localhost:3080/v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/cloud/nodes/b354a67c-f1a9-4e85-927f-2677ec8a8fce/adapters/0/ports/5/nio {'content-type': 'application/json'}
2020-03-26 15:59:38 INFO cloud.py:399 Cloud "Cloud1" [b354a67c-f1a9-4e85-927f-2677ec8a8fce]: NIO NIO UDP bound to port 5
2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:204 sending bridge create b354a67c-f1a9-4e85-927f-2677ec8a8fce-5

2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:273 returned result ["bridge 'b354a67c-f1a9-4e85-927f-2677ec8a8fce-5' created"]
2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:204 sending bridge add_nio_udp b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 10001 127.0.0.1 10000

2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:273 returned result ["NIO UDP added to bridge 'b354a67c-f1a9-4e85-927f-2677ec8a8fce-5'"]
2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:204 sending bridge reset_packet_filters b354a67c-f1a9-4e85-927f-2677ec8a8fce-5

2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:273 returned result []
2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:204 sending bridge add_nio_tap b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 "tap0"

2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:273 returned result ["NIO TAP added to bridge 'b354a67c-f1a9-4e85-927f-2677ec8a8fce-5'"]
2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:204 sending bridge start b354a67c-f1a9-4e85-927f-2677ec8a8fce-5

2020-03-26 15:59:38 DEBUG ubridge_hypervisor.py:273 returned result ["bridge 'b354a67c-f1a9-4e85-927f-2677ec8a8fce-5' started"]
2020-03-26 15:59:38 INFO response.py:58 POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/cloud/nodes/b354a67c-f1a9-4e85-927f-2677ec8a8fce/adapters/0/ports/5/nio
2020-03-26 15:59:38 DEBUG response.py:59 {'Host': 'localhost:3080', 'Content-Type': 'application/json', 'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Python/3.6 aiohttp/3.6.2', 'Authorization': 'Basic YWRtaW46Z3dydlFwekJqRXl4bFdiQ1NhTVhsQmR0R2pPbU9LQno3b0RrbmdhU3dGZGk1dEZFZndYbDZ2azAxUmdXMzNBNw==', 'Content-Length': '106'}
2020-03-26 15:59:38 DEBUG response.py:61 {'lport': 10001, 'rhost': '127.0.0.1', 'rport': 10000, 'type': 'nio_udp', 'filters': {}, 'suspend': False}
2020-03-26 15:59:38 INFO response.py:62 Response: 201 Created
2020-03-26 15:59:38 DEBUG response.py:63 {'Connection': 'close', 'X-Route': '/v2/compute/projects/{project_id}/cloud/nodes/{node_id}/adapters/{adapter_number:\\d+}/ports/{port_number:\\d+}/nio', 'Server': 'Python/3.6 GNS3/2.2.6', 'Content-Type': 'application/json'}
2020-03-26 15:59:38 DEBUG response.py:65 {'lport': 10001, 'rhost': '127.0.0.1', 'rport': 10000, 'type': 'nio_udp'}
2020-03-26 15:59:38 INFO web_log.py:233 127.0.0.1 [26/Mar/2020:14:59:38 +0000] "POST /v2/compute/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/cloud/nodes/b354a67c-f1a9-4e85-927f-2677ec8a8fce/adapters/0/ports/5/nio HTTP/1.1" 201 378 "-" "Python/3.6 aiohttp/3.6.2"
2020-03-26 15:59:38 DEBUG project.py:1059 Write /home/bizzi/GNS3/projects/untitled/untitled.gns3
2020-03-26 15:59:38 INFO response.py:58 POST /v2/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/links
2020-03-26 15:59:38 DEBUG response.py:59 {'Authorization': 'Basic YWRtaW46Z3dydlFwekJqRXl4bFdiQ1NhTVhsQmR0R2pPbU9LQno3b0RrbmdhU3dGZGk1dEZFZndYbDZ2azAxUmdXMzNBNw==', 'User-Agent': 'GNS3 QT Client v2.2.6', 'Content-Type': 'application/json', 'Content-Length': '228', 'Connection': 'Keep-Alive', 'Accept-Encoding': 'gzip, deflate', 'Accept-Language': 'it-IT,en,*', 'Host': 'localhost:3080'}
2020-03-26 15:59:38 DEBUG response.py:61 {'nodes': [{'node_id': '18f5c42d-80f6-4fac-9886-242632e81901', 'adapter_number': 0, 'port_number': 0}, {'node_id': 'b354a67c-f1a9-4e85-927f-2677ec8a8fce', 'adapter_number': 0, 'port_number': 5}], 'filters': {}, 'suspend': False}
2020-03-26 15:59:38 INFO response.py:62 Response: 201 Created
2020-03-26 15:59:38 DEBUG response.py:63 {'Connection': 'close', 'X-Route': '/v2/projects/{project_id}/links', 'Server': 'Python/3.6 GNS3/2.2.6', 'Content-Type': 'application/json'}
2020-03-26 15:59:38 DEBUG response.py:65 {'capture_compute_id': None, 'capture_file_name': None, 'capture_file_path': None, 'capturing': False, 'filters': {}, 'link_id': '41a58ef8-a7e5-4921-81c2-16e4c723a0f8', 'link_type': 'ethernet', 'nodes': [{'adapter_number': 0, 'label': {'style': 'font-family: TypeWriter;font-size: 10.0;font-weight: bold;fill: #000000;fill-opacity: 1.0;', 'text': '0/0'}, 'node_id': '18f5c42d-80f6-4fac-9886-242632e81901', 'port_number': 0}, {'adapter_number': 0, 'label': {'style': 'font-family: TypeWriter;font-size: 10.0;font-weight: bold;fill: #000000;fill-opacity: 1.0;', 'text': '0/5'}, 'node_id': 'b354a67c-f1a9-4e85-927f-2677ec8a8fce', 'port_number': 5}], 'project_id': '8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee', 'suspend': False}
2020-03-26 15:59:38 INFO web_log.py:233 ::1 [26/Mar/2020:14:59:37 +0000] "POST /v2/projects/8f161e98-c93c-4fa4-9e4e-b94e6a13e8ee/links HTTP/1.1" 201 1196 "-" "GNS3 QT Client v2.2.6"
2020-03-26 15:59:38 ERROR hypervisor.py:195 uBridge process has stopped, return code: 1
uBridge version 0.9.17 running with libpcap version 1.8.1
Hypervisor TCP control server started (IP localhost port 32863).
UDP tunnel connecting from local port 10001 to IPv4 address 127.0.0.1 on port 10000
Source NIO listener thread for b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 has started
Destination NIO listener thread for b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 has started
Received 110 bytes on bridge 'b354a67c-f1a9-4e85-927f-2677ec8a8fce-5' (destination NIO)
0000: 33 33 00 00 00 16 da d5 26 aa 7a 15 86 dd 60 00 33........z.....
0010: 00 00 00 38 00 01 fe 80 00 00 00 00 00 00 d8 d5 ...8............
0020: 26 ff fe aa 7a 15 ff 02 00 00 00 00 00 00 00 00 ....z...........
0030: 00 00 00 00 00 16 3a 00 05 02 00 00 01 00 8f 00 ................
0040: 79 a1 00 00 00 02 04 00 00 00 ff 02 00 00 00 00 y...............
0050: 00 00 00 00 00 00 00 00 00 fb 04 00 00 00 ff 02 ................
0060: 00 00 00 00 00 00 00 00 00 01 ff aa 7a 15       ............z.

recv: Connection refused
Source NIO listener thread for b354a67c-f1a9-4e85-927f-2677ec8a8fce-5 has stopped because of an error: Invalid argument 


```

If I start manually the uBridge hypervisor on another shell and create/start the bridge after that the IOU process was started all works.

The host is an Ubuntu 18.04.4 LTS on a multiprocessor server.
The GNS3 was added from deb http://ppa.launchpad.net/gns3/ppa/ubuntu bionic main

The same project on a 2.2.5 works fine.

Can someone help me please. :)

Many thanks.

Fabio.

**Comment by procisco** [comment]
Hello, just want to add I have a similar setup and get the same errors. I do not use IOU but QEMU for VIRL images and Ubuntu guest hosts. The ubridge fails with the same error for links between VIRL  hosts. Dynamips images run okay. I have little experience debugging GNS3 as it's always been rock solid, so not sure how I can help other than confirming similar issues.

**Comment by grossmj** [comment]
We have released uBridge 0.9.18 which solves this problem. Please upgrade your system and things should work again.

Thanks!
